File: 12/pcHebras.cpp
#include <iostream>
#include <condition_variable>
#include <mutex>
#include <thread>
#include <random>
#include <cmath>
#include <atomic>
#include <chrono>
#include <ctime>
#include "pcMonitor.cpp"
using namespace std;

pcMonitor monitor(100);


void run(int i){
	if(i%2)monitor.producir(rand()%100);
	else monitor.consumir();

}

int main(){
	std::chrono::time_point<std::chrono::system_clock> start,end;
	start=std::chrono::system_clock::now();
	int tareas=4;//pasar por argumento
	thread hilos[10];
	srand(time(NULL));
	for(int i=0;i<10;++i){
		hilos[i]= thread(run,i);
	}

	for(int i=0;i<10;++i){
		hilos[i].join();
	}

	end=std::chrono::system_clock::now();
	std::chrono::duration<double> seconds= end-start;
	cout<<"Ha tardado: "<<seconds.count()<< "s\n";

}

File: 12/pcHebras.java
import java.util.*;
import java.util.concurrent.*;
//Modificar, cambiar a threadpool y separar el productor del consumidor
class pcHebras implements Runnable{

	private static int n,aciertos;
	private static double PI;
	private static Scanner sc = new Scanner(System.in);
	private static pcMonitor monitor new pcMonitor(100);


	public pcHebras(int n){
		this.n=n;
											}

	public void run(){

		if(n%2)producir();
		else consumir();

			}

	void producir(){
			monitor.insertar();//tienes que cambiar el monitor

						}

	 void consumir(){

			if(monitor.extraer()<=1)++aciertos;

						}

	public static void main(String[] args){
		long time_start, time_end;
		System.out.println("Cuantas inserciones y extracciones quieres?");
		int n=sc.nextInt();
		time_start = System.currentTimeMillis();
		//ThreadPoolExecutor pool = new ThreadPoolExecutor(4,4,0L,TimeUnit MILLISECONDS,new LinkedBlockingQueue<Runnable>());//cambiar el 4 para hacer el speedup
		Thread []hilos= new Thread[n];//cambiar
		for(int i=0;i<n*2;++i){
		//pool.execute(new Thread(new pcHebras(i)));
		hilos[i]= new Thread(new pcHebras(i));
		hilos[i].start();
							}

		for(int i=0;i<n*2;++i){
				hilos[i].join();
		}

		/**pool.shutdown();
				try{
						pool.awaitTermination(2, TimeUnit.SECONDS);
				}catch(Exception ex){}*/
	//	System.out.println("Aciertos= "+aciertos+"\nNumeros= "+(5*n));
		PI=(4*((double)(aciertos)/(double)(n*5)));
		System.out.println("La aproximacion del numero PI es "+PI);
		time_end = System.currentTimeMillis();
		System.out.println("the task has taken "+ ( time_end - time_start ) +" milliseconds");
										}


				}
